Redunant declarations compiler error for UnityPrintFail, UnityPrintOk and setUp

Hi,

I'm using gcc arm embedded toolchain (2015q4, gcc 5.2) and have enabled `-Werror=redundant-decls` compiler flag. It seems the functions `UnityPrintFail`, `UnityPrintOk` and `setUp` have been defined to many times.
The definitions on line 70 and 71 and 1264 are not necessary and redundant.

Here is the full log:
```
Unity/src/unity.c:303:6: error: redundant redeclaration of 'UnityPrintFail' [-Werror=redundant-decls]
 void UnityPrintFail(void);
      ^
Unity/src/unity.c:70:6: note: previous declaration of 'UnityPrintFail' was here
 void UnityPrintFail(void);
      ^
Unity/src/unity.c:309:6: error: redundant redeclaration of 'UnityPrintOk' [-Werror=redundant-decls]
 void UnityPrintOk(void);
      ^
Unity/src/unity.c:71:6: note: previous declaration of 'UnityPrintOk' was here
 void UnityPrintOk(void);
      ^
Unity/src/unity.c:1264:10: error: redundant redeclaration of 'setUp' [-Werror=redundant-decls]
     void setUp(void);
          ^
In file included from Unity/src/unity.c:7:0:
Unity/src/unity.h:18:6: note: previous declaration of 'setUp' was here
 void setUp(void);
      ^
Unity/src/unity.c:1265:10: error: redundant redeclaration of 'tearDown' [-Werror=redundant-decls]
     void tearDown(void);
          ^
In file included from Unity/src/unity.c:7:0:
Unity/src/unity.h:19:6: note: previous declaration of 'tearDown' was here
 void tearDown(void);
```

It seems when I add this flag to the tests makefile it seems to fail also. It is a good idea to add this flag to the tests makefile when this is fixed. Using (debian 8 amd64):

```
cc --version
cc (Debian 4.9.2-10) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
jjacobs@dev03:~/repos/github/Unity/test$ make
cd build && \
gcc -std=c99 -pedantic -Wall -Wextra -Werror -Wredundant-decls  -D UNITY_OUTPUT_CHAR=putcharSpy -D UNITY_INCLUDE_DOUBLE  ../../src/unity.c  ../tests/testunity.c  ../build/testunityRunner.c -fprofile-arcs -ftest-coverage -I ../../src -o ../build/testunity-cov.exe
../../src/unity.c:303:6: error: redundant redeclaration of ‘UnityPrintFail’ [-Werror=redundant-decls]
 void UnityPrintFail(void);
      ^
../../src/unity.c:70:6: note: previous declaration of ‘UnityPrintFail’ was here
 void UnityPrintFail(void);
      ^
../../src/unity.c:309:6: error: redundant redeclaration of ‘UnityPrintOk’ [-Werror=redundant-decls]
 void UnityPrintOk(void);
      ^
../../src/unity.c:71:6: note: previous declaration of ‘UnityPrintOk’ was here
 void UnityPrintOk(void);
      ^
../../src/unity.c:1264:10: error: redundant redeclaration of ‘setUp’ [-Werror=redundant-decls]
     void setUp(void);
          ^
In file included from ../../src/unity.c:7:0:
../../src/unity.h:18:6: note: previous declaration of ‘setUp’ was here
 void setUp(void);
      ^
../../src/unity.c:1265:10: error: redundant redeclaration of ‘tearDown’ [-Werror=redundant-decls]
     void tearDown(void);
          ^
In file included from ../../src/unity.c:7:0:
../../src/unity.h:19:6: note: previous declaration of ‘tearDown’ was here
 void tearDown(void);
      ^
cc1: all warnings being treated as errors
Makefile:21: recipe for target 'coverage' failed
make: *** [coverage] Error 1
```

--- a/src/unity.c
+++ b/src/unity.c
@@ -67,9 +67,6 @@ const _U_UINT UnitySizeMask[] =
 #endif
 };
 
-void UnityPrintFail(void);
-void UnityPrintOk(void);
-
 //-----------------------------------------------
 // Pretty Printers & Test Result Output Handlers
 //-----------------------------------------------
@@ -1261,8 +1258,6 @@ void UnityIgnore(const char* msg, const UNITY_LINE_TYPE line)
 
 //-----------------------------------------------
 #if defined(UNITY_WEAK_ATTRIBUTE)
-    void setUp(void);
-    void tearDown(void);
     UNITY_WEAK_ATTRIBUTE void setUp(void) { }
     UNITY_WEAK_ATTRIBUTE void tearDown(void) { }
 #elif defined(UNITY_WEAK_PRAGMA)
@@ -1270,9 +1265,6 @@ void UnityIgnore(const char* msg, const UNITY_LINE_TYPE line)
     void setUp(void) { }
 #   pragma weak tearDown
     void tearDown(void) { }
-#else
-    void setUp(void);
-    void tearDown(void);
 #endif
 //-----------------------------------------------
 void UnityDefaultTestRun(UnityTestFunction Func, const char* FuncName, const int FuncLineNum)

--- a/src/unity.c
+++ b/src/unity.c
@@ -7,6 +7,11 @@
 #include "unity.h"
 #include <stddef.h>
 
+#ifndef UNITY_OUTPUT_CHAR_USE_PUTC
+//If defined as something else, make sure we declare it here so it's ready for use
+extern int UNITY_OUTPUT_CHAR(int);
+#endif
+
 #define UNITY_FAIL_AND_BAIL   { Unity.CurrentTestFailed  = 1; longjmp(Unity.AbortFrame, 1); }
 #define UNITY_IGNORE_AND_BAIL { Unity.CurrentTestIgnored = 1; longjmp(Unity.AbortFrame, 1); }
 /// return prematurely if we are already in failure or ignore state

--- a/src/unity_internals.h
+++ b/src/unity_internals.h
@@ -291,9 +291,12 @@ typedef UNITY_DOUBLE_TYPE _UD;
 //Default to using putchar, which is defined in stdio.h
 #include <stdio.h>
 #define UNITY_OUTPUT_CHAR(a) putchar(a)
-#else
-//If defined as something else, make sure we declare it here so it's ready for use
-extern int UNITY_OUTPUT_CHAR(int);
+// We need to flag the output char function uses putc in
+//  unity.c the extern function is not declared then.
+// Previously the extern was declared in this header but
+//  when redundant function declaration compiler flag is enabled
+//  it wont compile.
+#define UNITY_OUTPUT_CHAR_USE_PUTC
 #endif
 
 #ifndef UNITY_PRINT_EOL

--- a/test/Makefile
+++ b/test/Makefile
@@ -3,7 +3,7 @@ ifeq ($(shell uname -s), Darwin)
 CC = clang
 endif
 #DEBUG = -O0 -g
-CFLAGS += -std=c99 -pedantic -Wall -Wextra -Werror
+CFLAGS += -std=c99 -pedantic -Wall -Wextra -Werror -Wredundant-decls
 CFLAGS += $(DEBUG)
 DEFINES = -D UNITY_OUTPUT_CHAR=putcharSpy -D UNITY_INCLUDE_DOUBLE
 SRC = ../src/unity.c tests/testunity.c build/testunityRunner.c

--- a/test/tests/testunity.c
+++ b/test/tests/testunity.c
@@ -8,6 +8,8 @@
 #include "unity.h"
 #include <string.h>
 
+int putcharSpy(int c);
+
 // Dividing by these constants produces +/- infinity.
 // The rationale is given in UnityAssertFloatIsInf's body.
 #ifndef UNITY_EXCLUDE_FLOAT

